packages/blocks/src/page-block/utils/cursor.ts

```diff
@@ -76,55 +78,106 @@ export function getDragDirection(e: SelectionEvent): DragDirection {
-  const isStartAnchor = ['left-top', 'right-top'].includes(direction);
+  return { selectedType, direction, anchor: curRange };
+}
+
+export function calcPositionPointByRange(
+  range: Range,
+  direction: DragDirection
+) {
+  const { startContainer, startOffset, endContainer, endOffset } = range;
+
+  const isStartAnchor = direction.includes('top');
   const [targetNode, offset] = isStartAnchor
-    ? [anchorNode, anchorOffset]
-    : [focusNode, focusOffset];
+    ? [startContainer, startOffset]
+    : [endContainer, endOffset];
+  const lineRange = new Range();
 
-  assertExists(targetNode);
-  const range = document.createRange();
-  range.setStart(targetNode, offset);
-  // Workaround for select to the start of line
-  // If not set end, the getBoundingClientRect of range will return the wrong value
-  //
-  // For example
-  // line1↩
-  //      ↑ range.getBoundingClientRect will return this position
-  //        if you select `line2` from back to front,
-  // line2↩
-  const endOffset = Math.min(offset + 1, targetNode.textContent?.length || 0);
-  if (isStartAnchor) {
-    range.setEnd(targetNode, endOffset);
+  if (direction.includes('center')) {
+    // Use whole range
+    lineRange.setStart(startContainer, startOffset);
+    lineRange.setEnd(endContainer, endOffset);
+  } else {
+    lineRange.setStart(targetNode, offset);
   }
 
   // XXX the workaround is very ugly, please improve after you find a way to identify the range is empty
   // Workaround select to empty line will get empty range
   // If range is empty, range.getBoundingClientRect of range will return the empty value({ x: 0, y: 0 ...})
-  const rangeRect = range.getBoundingClientRect();
+  const tmpRect = lineRange.getBoundingClientRect();
   const isWrongRect =
-    rangeRect.x === 0 &&
-    rangeRect.y === 0 &&
-    rangeRect.width === 0 &&
-    rangeRect.height === 0;
+    tmpRect.x === 0 &&
+    tmpRect.y === 0 &&
+    tmpRect.width === 0 &&
+    tmpRect.height === 0;
 
   if (isWrongRect) {
-    // Try to select something prevent range.getBoundingClientRect return empty value
-    range.setEnd(targetNode, endOffset);
+    // Fallback to use whole range
+    lineRange.setStart(startContainer, startOffset);
+    lineRange.setEnd(endContainer, endOffset);
   }
-  // You can uncomment follow line to debug the positioning range
-  // selection.removeAllRanges();
-  // selection.addRange(range);
-  return { selectedType, direction, anchor: range };
+  // resetNativeSelection(lineRange);
+
+  const lineRect = lineRange.getBoundingClientRect();
+  const isBottom = direction.includes('bottom');
+  const positioningPoint = {
+    // Use the center of the position rect
+    x: lineRect.x + lineRect.width / 2,
+    // if the direction is bottom, use the bottom of the position rect
+    y: lineRect.y + (isBottom ? lineRect.height : 0),
+  };
+  return positioningPoint;
 }
+
+export const calcSafeCoordinate = ({
```

Comment:
Can we make this a `function` and its params an `interface` or `type`? This style looks a bit too terse.